ApplicationLibCode/ReservoirDataModel/RigMswCenterLineCalculator.cpp

```diff
@@ -0,0 +1,331 @@
+
+    std::vector<WellBranch> wellBranches = mergeShortBranchesIntoLongBranches( resultBranches );
+
+    // Connect outlet segment of branches to parent branch
+
+    for ( const auto& resultBranch : resultBranches )
+    {
+        if ( resultBranch.m_branchResultPoints.empty() ) continue;
+
+        const auto firstResultPoint = resultBranch.m_branchResultPoints.front();
+
+        for ( auto& wellBranch : wellBranches )
+        {
+            if ( wellBranch.m_branchId == resultBranch.m_ertBranchId )
+            {
+                if ( firstResultPoint.branchId() == resultBranch.m_ertBranchId )
+                {
+                    // The first result point is on the same branch, use well head as outlet
+                    RigWellResultPoint outletResultPoint = wellFrame.m_wellHead;
+
+                    auto gridAndCellIndex = std::make_pair( outletResultPoint.gridIndex(), outletResultPoint.cellIndex() );
+                    wellBranch.m_segmentsWithGridCells[outletResultPoint.segmentId()].push_back( gridAndCellIndex );
+                }
+                else
+                {
+                    // The first result point on a different branch. Find the branch and add the grid cell
+                    for ( const auto& candidateResultBranch : wellBranches )
+                    {
+                        if ( firstResultPoint.branchId() == candidateResultBranch.m_branchId )
+                        {
+                            std::pair<size_t, size_t> gridAndCellIndexForTieIn;
+
+                            for ( const auto& [segment, gridAndCellIndices] : candidateResultBranch.m_segmentsWithGridCells )
+                            {
+                                if ( segment > firstResultPoint.segmentId() ) continue;
+                                if ( !gridAndCellIndices.empty() )
+                                {
+                                    gridAndCellIndexForTieIn = gridAndCellIndices.front();
+                                }
+                            }
+
+                            wellBranch.m_segmentsWithGridCells[firstResultPoint.segmentId()].push_back( gridAndCellIndexForTieIn );
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    std::vector<SimulationWellCellBranch> simWellBranches;
+    for ( const auto& wellBranch : wellBranches )
+    {
+        std::vector<cvf::Vec3d>         cellCenterCoords;
+        std::vector<RigWellResultPoint> cellCenterResultPoints;
+
+        if ( wellBranch.m_branchId == 1 && !wellBranch.m_segmentsWithGridCells.empty() )
+        {
+            const auto& [firstSegment, gridAndCellIndices] = *wellBranch.m_segmentsWithGridCells.begin();
+            if ( !gridAndCellIndices.empty() )
+            {
+                const auto& [gridIndex, cellIndex] = gridAndCellIndices.front();
+
+                const RigCell& cell       = eclipseCaseData->grid( gridIndex )->cell( cellIndex );
```

Comment:
Can `gridIndex` be invalid for `eclipseCaseData`? If so - consider guarding with `eclipseCaseData->gridCount`?